flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -657,20 +657,25 @@ protected void cleanUpInvoke() throws Exception {
 	}
 
 	@Override
-	public final void cancel() throws Exception {
+	public final CompletableFuture<Void> cancel() {
 		isRunning = false;
 		canceled = true;
 
 		// the "cancel task" call must come first, but the cancelables must be
 		// closed no matter what
-		try {
-			cancelTask();
-		}
-		finally {
-			mailboxProcessor.allActionsCompleted();
-			cancelables.close();
+		return FutureUtils
+			.completedVoidFuture()
+			.thenRun(this::cancelTask)
```

Comment:
Because `cancelTask` can also throw an exception.
Leaving client code to deal with both method throwing an exception and Future completed exceptionally isn't very nice IMO.

Having a `finally` block (I guess that's what you mean) would make calling `mailboxProcessor.allActionsCompleted` more complicated (I think it should be called after future completes if it was created successfully).

This is btw an idiomatic way in some languages like Scala.

Reply:
> Leaving client code to deal with both method throwing an exception and Future completed exceptionally isn't very nice IMO.

That would be an argument if not for the fact that exceptions are (and still should be probably) handled via the caller anyway.

But I guess you are right. To avoid code duplication with try/catch block, you would have to extract the `whenComplete(...)` content to a separate method. This could be argued which is a better/easier to understand way, so if you want let's leave it as it is.